sync/metrics.go

```diff
@@ -11,40 +12,152 @@ import (
+		"hdr_total_synced_headers",
 		metric.WithDescription("total synced headers"),
 	)
 	if err != nil {
 		return nil, err
 	}
 
-	m := &metrics{
-		totalSyncedGauge: totalSynced,
+	syncLoopStarted, err := meter.Int64Counter(
+		"hdr_sync_loop_started",
+		metric.WithDescription("sync loop started"),
+	)
+	if err != nil {
+		return nil, err
 	}
 
-	callback := func(ctx context.Context, observer metric.Observer) error {
-		observer.ObserveFloat64(totalSynced, float64(m.totalSynced.Load()))
-		return nil
+	trustedPeersOutOfSync, err := meter.Int64Counter(
+		"hdr_tr_peers_out_of_sync",
+		metric.WithDescription("trusted peers out of sync"),
+	)
+	if err != nil {
+		return nil, err
 	}
-	_, err = meter.RegisterCallback(callback, totalSynced)
+
+	laggingHeadersStart, err := meter.Int64Counter(
+		"hdr_sync_lagging_hdr_start",
+		metric.WithDescription("lagging header start"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	subjectiveHead, err := meter.Int64ObservableGauge(
+		"hdr_sync_subjective_head",
+		metric.WithDescription("subjective head height"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	blockTime, err := meter.Float64Histogram(
+		"hdr_sync_actual_blockTime_ts",
+		metric.WithDescription("duration between creation of 2 blocks"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &metrics{
+		totalSyncedInst:       totalSynced,
+		syncLoopStarted:       syncLoopStarted,
+		trustedPeersOutOfSync: trustedPeersOutOfSync,
+		laggingHeadersStart:   laggingHeadersStart,
+		blockTime:             blockTime,
+		subjectiveHeadInst:    subjectiveHead,
+		headersThreshold:      headersThreshold,
+	}
+
+	m.syncReg, err = meter.RegisterCallback(m.observeMetrics, m.totalSyncedInst, m.subjectiveHeadInst)
 	if err != nil {
 		return nil, err
 	}
 
 	return m, nil
 }
 
-// recordTotalSynced records the total amount of synced headers.
+func (m *metrics) observeMetrics(ctx context.Context, obs metric.Observer) error {
+	err := m.observeTotalSynced(ctx, obs)
+	if err != nil {
+		return err
+	}
+
+	return m.observeNewHead(ctx, obs)
+}
+
+func (m *metrics) observeTotalSynced(_ context.Context, obs metric.Observer) error {
+	obs.ObserveInt64(m.totalSyncedInst, m.totalSynced.Load())
+	return nil
+}
+
+func (m *metrics) observeNewHead(_ context.Context, obs metric.Observer) error {
+	obs.ObserveInt64(m.subjectiveHeadInst, m.subjectiveHead.Load())
+	return nil
+}
+
 func (m *metrics) recordTotalSynced(totalSynced int) {
+	m.observe(context.Background(), func(_ context.Context) {
+		m.totalSynced.Add(int64(totalSynced))
+	})
+}
+
+func (m *metrics) recordSyncLoopStarted(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.syncLoopStarted.Add(ctx, 1)
+	})
+}
+
+func (m *metrics) recordTrustedPeersOutOfSync(ctx context.Context) {
+	m.observe(ctx, func(ctx context.Context) {
+		m.trustedPeersOutOfSync.Add(ctx, 1)
+	})
+}
+
+func (m *metrics) observeNewSubjectiveHead(ctx context.Context, height int64, timestamp time.Time) {
```

Comment:
A bit of naming inconsistency between record and observe

Reply:
I would honestly go without both. Having e.g.
`s.metrics.newSubjectiveHead` is pretty readable.